keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,478 @@
+response is small (it fits a single page, assuming you won't make it extremely
+small), whereas the number of objects to process is proportional to cluster-size
+(so fairly big). For example, when kubelets requests pods schedule against it in
+a 5k node cluster with 30pods/node, the kube-apiserver must list the 150k pods
+from etcd and then filter that list down to the list of 30 pods that the kubelet
+actually need. This must occur for each list request from each of the 5k
+kubelets. If served from watch cache, this same request can be served by simply
+filtering out the 30 pods each kubelet needs from the data in the cache.
+
+In addition to the improvements to scale and performance, we aim to resolve a
+specific problem. The long standing "stale read" issue
+(https://github.com/kubernetes/kubernetes/issues/59848) remains open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted and this issue
+would be solved. "Going back in time" can curently happen if the initial list
+request is served from a stale watch cache with data much older than the
+reflector has previously observed or if the api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scalability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+### Non-Goals
+
+<<[UNRESOLVED @deads]>>
+- Avoid allowing true quorum reads. We should think carefully about this, see: https://github.com/kubernetes/enhancements/pull/1404#discussion_r381528406
+<<[/UNRESOLVED]>>
+
+## Proposal
+
+### Consistent reads from cache
+
+Guard this by a `WatchCacheConsistentReads` feature gate.
+
+#### Use WithProgressNotify to enable automatic watch updates
+
+Create etcd watches with `WithProgressNotify` enabled (available in all etcd 3.x versions).
+
+When `WithProgressNotify` is enabled on an etcd watch, etcd sends progress
+events to the watch automatically. By default etcd sends progress events every
+10 minutes, which is not frequent enough to be useful for our needs, so we will
+modify etcd to send them more frequently.
+
+When an consistent LIST request is received and the watch cache is enabled:
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
```

Comment:
What does "current revision" mean exactly? We can get the largest modified revision for the listed resources (vs current revision of the entire key space). This can reduce the block time (for less frequently changed objects, there should be just no blocking).

Alternatively, we can add a new watch request type "urgent" or something to let etcd simply deliver all pending watch events or an empty result (if there is no new changes).